### PR TITLE
Add script to reconstruct points from transactions

### PR DESCRIPTION
**Pull request by motechFR**

### CODE REVIEW

File: packages/scoutgame/src/points/sendPoints.ts [hunk 1]
```diff
@@ -1,33 +1,44 @@
...
 
 export async function sendPoints({
   builderId,
   season = currentSeason,
   week = getCurrentWeek(),
   points,
+  description = 'Bonus points',
   earnedAsBuilder = false,
-  hideFromNotifications = false
+  claimed = true,
```
Comment (mattcasey) [inline packages/scoutgame/src/points/sendPoints.ts:17]: claimed should be a required field, no default please. that could be unexpected

File: packages/scoutgame/src/points/sendPoints.ts [hunk 2]
```diff
@@ -1,33 +1,44 @@
...
+
+import { refreshPointStatsFromHistory } from './refreshPointStatsFromHistory';
+import { incrementPointsEarnedStats } from './updatePointsEarned';
 
 export async function sendPoints({
   builderId,
   season = currentSeason,
   week = getCurrentWeek(),
   points,
+  description = 'Bonus points',
```
Comment (mattcasey) [inline packages/scoutgame/src/points/sendPoints.ts:15]: no default for description please. i dont know what "Bonus points" means, is this something chris said to use?

File: packages/scoutgame/src/points/updatePointsEarned.ts [hunk 3]
```diff
@@ -64,3 +67,71 @@ export async function updatePointsEarned({
...
+      }
+    });
+
+    return {
+      seasonStats,
+      allTimeStats
+    };
+  }
+
+  if (tx) {
```
[the remark this answers is not in the document]
Comment (motechFR) [reply packages/scoutgame/src/points/updatePointsEarned.ts:140]: Thank you for spotting this, fixed

File: packages/scoutgame/src/points/sendPoints.ts [hunk 4]
```diff
@@ -36,45 +47,66 @@ export async function sendPoints({
...
+  }
+
+  if (tx) {
+    return txHandler(tx);
+  } else {
+    return prisma.$transaction(txHandler);
+  }
+}
+
+const fids: number[] = [
```
Comment (mattcasey) [inline packages/scoutgame/src/points/sendPoints.ts:83]: this belongs in a script someplace. this file is just meant to be a utility method to be imported?

File: packages/scoutgame/src/points/sendPoints.ts [hunk 5]
```diff
@@ -36,45 +47,66 @@ export async function sendPoints({
...
+    log.info(`Issuing points to fid: ${fid}`);
+
+    const scout = await prisma.scout.findFirstOrThrow({
+      where: {
+        farcasterId: fid
+      }
+    });
+
+    await prisma.$transaction(async (tx) => {
+      await sendPoints({
```
Comment (mattcasey) [inline packages/scoutgame/src/points/sendPoints.ts:98]: did you mean to pass in the tx?

File: packages/scoutgame/src/points/sendPoints.ts [hunk 6]
```diff
@@ -36,45 +47,66 @@ export async function sendPoints({
...
-          })
-        ]
-      : [])
-  ]);
+    });
+    await incrementPointsEarnedStats({
+      season,
+      userId: builderId,
+      builderPoints: earnedAsBuilder ? points : 0,
+      scoutPoints: !earnedAsBuilder ? points : 0,
```
Comment (mattcasey) [inline packages/scoutgame/src/points/sendPoints.ts:71]: these points should not be 'earned as' at all. it seems like you'd be counting these as earned as a scout

File: packages/scoutgame/src/points/sendPoints.ts (hunk 1 above)
Comment (motechFR) [reply packages/scoutgame/src/points/sendPoints.ts:17]: fixed

File: packages/scoutgame/src/points/sendPoints.ts (hunk 2 above)
Comment (motechFR) [reply packages/scoutgame/src/points/sendPoints.ts:15]: Made this a mandatory field also as with claimed flag

File: packages/scoutgame/src/points/refreshPointStatsFromHistory.ts [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+      tx.pointsReceipt.findMany({
+        where: {
+          senderId: userId,
+          event: {
+            season
+          }
+        }
+      }),
+      // Points received as builder
+      tx.pointsReceipt.findMany({
```
Comment (mattcasey) [inline packages/scoutgame/src/points/refreshPointStatsFromHistory.ts:62]: why not just look by receipientId? what cases are you filtering out with the OR clause?

File: packages/scoutgame/src/points/refreshPointStatsFromHistory.ts [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+        .then((user) => user.id);
+
+  const [pointsSpentRecords, pointsReceivedAsBuilderRecords, pointsReceivedAsScoutRecords, allPointsReceivedRecords] =
+    await Promise.all([
+      // Points spent
+      tx.pointsReceipt.findMany({
+        where: {
+          senderId: userId,
+          event: {
+            season
```
Comment (mattcasey) [inline packages/scoutgame/src/points/refreshPointStatsFromHistory.ts:57]: points are cross-season, why are we filtering by season?

File: packages/scoutgame/src/points/sendPoints.ts (hunk 4 above)
Comment (motechFR) [reply packages/scoutgame/src/points/sendPoints.ts:83]: Thank you, moved this to /scripts under scoutgamecron

File: packages/scoutgame/src/points/sendPoints.ts (hunk 5 above)
Comment (motechFR) [reply packages/scoutgame/src/points/sendPoints.ts:98]: fixed

File: packages/scoutgame/src/points/refreshPointStatsFromHistory.ts [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+  const pointsSpent = pointsSpentRecords.reduce((acc, { value }) => acc + value, 0);
+
+  const pointsReceivedAsBuilder = pointsReceivedAsBuilderRecords.reduce((acc, { value }) => acc + value, 0);
+  const pointsReceivedAsScout = pointsReceivedAsScoutRecords.reduce((acc, { value }) => acc + value, 0);
+
+  const allPointsReceived = allPointsReceivedRecords.reduce((acc, { value }) => acc + value, 0);
+
+  const balance = pointsReceivedAsBuilder + pointsReceivedAsScout - pointsSpent;
+
+  const missingPointRecords = allPointsReceivedRecords.filter(
```
Comment (mattcasey) [inline packages/scoutgame/src/points/refreshPointStatsFromHistory.ts:161]: i guess this was for debugging but its not used

**Pull request merged**